crates/resolver/src/cache.rs

```diff
@@ -135,12 +135,64 @@ impl Entry {
             }
             Err(e) => {
                 let mut e = e.clone();
+
+                // The NoRecords error may contain up to four fields with TTL values present: negative_ttl, soa, authorities, and ns.
+                // For completeness, we update each field, if present.
                 if let NetErrorKind::Dns(DnsError::NoRecordsFound(NoRecords {
-                    negative_ttl: Some(ttl),
+                    negative_ttl,
+                    soa,
+                    authorities,
+                    ns,
                     ..
                 })) = &mut e.kind
                 {
-                    *ttl = ttl.saturating_sub(elapsed);
+                    if let Some(ttl) = negative_ttl {
+                        *ttl = ttl.saturating_sub(elapsed);
+                    }
+
+                    if let Some(soa) = soa {
+                        soa.decrement_ttl(elapsed);
+                    }
+
+                    if let Some(recs) = authorities {
+                        let updated_auths = recs
```

Comment:
Nit: suggest inlining the binding to `updated_auths` here (and `updated_ns`), below, to just call `authorities.replace(Arc::from(...collect::<Vec<_>>()))` directly (I think `Arc::from()` will [take care](https://doc.rust-lang.org/std/sync/struct.Arc.html#impl-From%3CVec%3CT,+A%3E%3E-for-Arc%3C%5BT%5D,+A%3E) of converting the `Vec` into a boxed slice).